draft-ietf-mls-protocol.md

```diff
@@ -1384,10 +1384,23 @@ enum {
     (255)
 } ContentType;
 
+enum {
+    invalid(0),
+    member(1),
+    external_authority(2),
+    external_self(3),
+    (255)
+} SenderType;
+
+struct {
+    SenderType sender_type;
+    uint32 sender;
+} Sender;
+
 struct {
     opaque group_id<0..255>;
     uint32 epoch;
-    uint32 sender;
+    Sender sender;
```

Comment:
There's no need to use the full `Sender` struct here, since you have to be a member to send an `MLSCiphertext`; just use uint32.  In other words, encrypting an MLSPlaintext with `Sender.sender_type != member` is an error, and an MLSPlaintext decrypted from an MLSCiphertext always has `Sender.sender_type != member`.

Reply:
As a bonus, this makes it syntactically impossible for an external sender to send an MLSCiphertext.

Reply:
iiuc - the proposal is to treat all MLSPlaintexts as `member` sender types? That would mean external senders would need some other mechanism for sender? How can we differentiate between self add and a preconfigured sender?

I'm confused by this comment - were you thinking this was in MLSSenderData?

Reply:
There's a difference between MLSPlaintext and MLSCiphertext here.  An MLSPlaintext can be constructed and sent by anyone; all you need is a signing key that is known to the group.  In order to construct an MLSCiphertext, you have to be a member of the group, since otherwise you won't have the right symmetric keys.  In other words, external senders can only send *unencrypted* messages to the group right now.  (I think Raphael is working on a PR to enable encryption to the group from outside.)

Reply:
Yea - I think we are on the same page. MLSPlaintext can be sent by anyone (in or outside the group) so it needs a full Sender object to disambiguate. However MLSCiphertext needs only the sender index, which is done by keeping MLSSenderData as is.

Reply:
Oh I see.  I was mistaken about what line we were looking at.  LGTM.